### PR TITLE
Fix #1296

### LINKED ISSUE #1296: blueman-applet occasionally launches 2 instances of blueman-tray
blueman: 2.1.3
BlueZ: 5.54
Distribution: Manjaro 20.0.3
Desktop environment: i3

I think this is a race condition involving the lock file.
If the `StatusIcon` and `AppIndicator` plugins are both enabled, then during startup the `StatusIcon` plugin will launch `blueman-tray` twice in `_on_plugins_changed()`: once when `StatusIcon` is loaded, once when `AppIndicator` is loaded. The function checks if there is an instance already running by reading the tray's lock file, but that's racy, since the tray can be running without having created the lock file. There's yet another check in `blueman-tray` using `check_single_instance()`, but that's racy too, here's how both trays can pass the check, assuming a stale lock file exists:
1. Tray 1 sees that a lock file exists, then reads a stale PID.
2. Tray 2 sees that a lock file exists, reads the same stale PID, removes the lock file, and writes its own lock file.
3. Tray 1 still thinks that the lock file is stale, so it removes tray 2's lock file and writes its own lock file.

Overall, there seem to be two problems here, and both need to be addressed for this to work correctly:
1. The check for a running tray in `_on_plugins_changed()` inside `StatusIcon`: even if `check_single_instance()` worked correctly, we could end up with the wrong instance of the tray running (i.e. the one launched first). This could be fixed by storing the PID of the tray process in a field of the `StatusIcon` class instead of reading it from the lock file, and reading the lock file when initializing that field in `StatusIcons`'s `__init__()` to handle the case where `blueman-tray` is launched first.
2. `check_single_instance()`: I think it can be fixed by serializing calls that check the same lock file. We can use [fcntl.flock](https://docs.python.org/3/library/fcntl.html#fcntl.flock) for that.

If my explanation is correct, then a workaround would be to disable the `AppIndicator` plugin on systems that can display GTK status icons.

### PR DESCRIPTION
**Pull request by j-piecuch**
This is my attempt at fixing #1296. It does two things:
- Serializes accesses to the same lock file from different processes using `fcntl.flock`
- Stores the PID of the tray application in the `StatusIcon` class

I'm open to feedback, especially since I don't have a lot of experience writing GTK/Python code.

### CODE REVIEW

File: blueman/Functions.py [hunk 1]
```diff
@@ -112,18 +110,14 @@ def launch(
...
-    else:
-        timestamp = gtktimestamp
-        display = Gdk.Display.get_default()
-        context = display.get_app_launch_context()
-        context.set_timestamp(timestamp)
+
+    timestamp = Gtk.get_current_event_time()
+    display = Gdk.Display.get_default()
+    context = display.get_app_launch_context()
+    context.set_timestamp(timestamp)
```
Comment (infirit) [inline blueman/Functions.py:120]: Don't undo this, see #1269 for details

Comment (j-piecuch) [reply blueman/Functions.py:127]: I see, I have adjusted the code to make up a timestamp when GTK can't provide one. I have tested the behavior from #1269 on MATE 1.24 and it seems to work fine.

**Comment by infirit** [comment]
Ok, apologies for you doing all this work but I already had something I was working on from 2 years ago. See #1304

With Gtk/Gio.Application we can handle uniqueness properly and there is no need anymore for check_single_instance.

**Comment by j-piecuch** [comment]
No problem, `Gtk.Application` looks like a much cleaner solution. Closing.

**Pull request closed without merging**